libraries/chain/controller.cpp

```diff
@@ -1793,11 +1793,12 @@ struct controller_impl {
             db.modify( pso, [&]( auto& ps ) {
                ps.preactivated_protocol_features.clear();
 
-               ps.activated_protocol_features.reserve( ps.activated_protocol_features.size()
-                                                         + new_protocol_feature_activations.size() );
-               for( const auto& feature_digest : new_protocol_feature_activations ) {
-                  ps.activated_protocol_features.emplace_back( feature_digest, pbhs.block_num );
-               }
+               std::size_t current_size = ps.activated_protocol_features.size();
+               std::size_t new_size     = current_size + new_protocol_feature_activations.size();
+               
+               ps.activated_protocol_features.clear_and_construct(new_size, current_size, [&](auto* dest, std::size_t idx) {
```

Comment:
I added `emplace_back` a while back, but I can't have an `append` of multiple items, because we need each value to be constructed in-place so the correct allocator is used.

Reply:
I was thinking the append could take a `const std::vector<T>&` and do what you have here.

Reply:
I'd have to construct a vector of `activated_protocol_feature` just to pass it to `append`, now we construct the `activated_protocol_feature` in-place. 
I could also `emplace_back` the new protocol features. I assume that in most cases there is only one, right?

Reply:
I was missing that `new_protocol_feature_activations` is not the same type as `preactivated_protocol_features`; was missing the `pbhs.block_num`. 

I do think a for loop of `emplace` would be much cleaner. Would be nice if `clear_and_construct` was private.

Reply:
I'll switch to a ` for loop of emplace`, but that would be very inefficient for many items.

Reply:
done.